foreign/java/external-processors/iggy-connector-flink/iggy-connector-library/src/main/java/org/apache/iggy/connector/serialization/JsonSerializationSchema.java

```
@@ -0,0 +1,163 @@
+     * @return new JsonSerializationSchema instance
+     */
+    public static <T> JsonSerializationSchema<T> withPartitionKeyExtractor(
+            Function<T, Integer> partitionKeyExtractor) {
+        return new JsonSerializationSchema<>(partitionKeyExtractor);
+    }
+
+    /**
+     * Creates a new JSON serializer with a custom ObjectMapper.
+     *
+     * @param objectMapper the Jackson ObjectMapper to use
+     */
+    public JsonSerializationSchema(ObjectMapper objectMapper) {
+        this(objectMapper, null);
+    }
+
+    /**
+     * Creates a new JSON serializer with custom ObjectMapper and partition key extractor.
+     *
+     * @param objectMapper the Jackson ObjectMapper to use
+     * @param partitionKeyExtractor function to extract partition key from elements
+     */
+    public JsonSerializationSchema(
+            ObjectMapper objectMapper,
+            Function<T, Integer> partitionKeyExtractor) {
+        if (objectMapper == null) {
+            throw new IllegalArgumentException("objectMapper cannot be null");
+        }
+        this.objectMapper = objectMapper;
+        this.partitionKeyExtractor = partitionKeyExtractor;
+    }
+
+    @Override
+    public byte[] serialize(T element) throws IOException {
+        if (element == null) {
+            return null;
+        }
+
+        try {
+            return getObjectMapper().writeValueAsBytes(element);
+        } catch (IOException e) {
+            throw new IOException(
+                    "Failed to serialize object of type " + element.getClass().getName(), e);
+        }
+    }
+
+    @Override
+    public Optional<Integer> extractPartitionKey(T element) {
+        if (partitionKeyExtractor != null && element != null) {
+            try {
+                Integer key = partitionKeyExtractor.apply(element);
+                return Optional.ofNullable(key);
+            } catch (Exception e) {
+                // Log and return empty if extraction fails
```

Please add the log statement. Also change the exception type to `RuntimeException`. `Exception` might silently catch a checked exception if the implementation changes in the future.